scipy/optimize/_numdiff.py

```diff
@@ -0,0 +1,297 @@
+
+    Returns
+    -------
+    h_adjusted : ndarray
+        Adjusted steps to use in derivative estimation.
+    """
+    lb, ub = bounds
+    if ub is None:
+        return h
+    if lb is None:
+        return -h
+
+    lb = np.asarray(lb)
+    ub = np.asarray(ub)
+
+    lower_dist = x0 - lb
+    upper_dist = ub - x0
+    forward = upper_dist > lower_dist
+
+    h_adjusted = np.empty_like(h)
+    h_adjusted[forward] = np.minimum(upper_dist[forward], h[forward])
+    h_adjusted[~forward] = -np.minimum(lower_dist[~forward], h[~forward])
+
+    return h_adjusted
+
+
+def _adjust_step_2nd_order(x0, h, bounds):
+    """Select a finite difference scheme and adjust the step size for the
+    first derivative estimation with the second order accuracy, such that
+    perturbed points are inside the bounds.
+
+    This function chooses between central finite difference and 3-point
+    forward (backward) finite difference. Both methods provide second order
+    accuracy.Central difference scheme is chosen with higher priority.
+
+    Parameters
+    ----------
+    x0 : ndarray
+        Initial point. Assumed to lie inside the bounds, no checks of that
+        are done.
+    h : ndarray
+        Desired positive step sizes. Might be decreased in extreme cases.
+    bounds : tuple of array_like
+        Lower and upper bounds on the variables. None means no bounds.
+
+    Returns
+    -------
+    h_adjusted : ndarray
+        Adjusted steps to use in derivative estimation.
+    scheme : ndarray
+        Array containing 0 and 1. If 0 then use central finite difference,
+        if 1 then use 3-point forward finite difference.
+    """
+    lb, ub = bounds
+    if lb is None and ub is None:
+        return h, np.zeros_like(h, dtype=int)
+
+    if lb is None:
+        lb = np.empty_like(h)
+        lb.fill(-np.inf)
+
+    if ub is None:
+        ub = np.empty_like(h)
+        ub.fill(np.inf)
+
+    scheme = np.zeros_like(h, dtype=int)
+    h_adjusted = h.copy()
+
+    lower_dist = x0 - lb
+    upper_dist = ub - x0
+
+    central = (lower_dist >= h) & (upper_dist >= h)
+
+    forward = ~central & (upper_dist >= 2 * h)
+    scheme[forward] = 1
+    h_adjusted[forward] = h[forward]
+
+    backward = ~central & (lower_dist >= 2 * h)
+    scheme[backward] = 1
+    h_adjusted[backward] = -h[backward]
+
+    central_adjusted = ~central & ~backward & ~forward
+    h_adjusted[central_adjusted] = np.minimum(lower_dist[central_adjusted],
+                                              upper_dist[central_adjusted])
+    return h_adjusted, scheme
+
+
+def _approx_forward(f, x0, h, f0, args, bounds):
+    h = _adjust_step_1st_order(x0, h, bounds)
+
+    if f0 is None:
+        f0 = f(x0, *args)
+
+    if x0.ndim == 0:
+        x = x0 + h
+        dx = x - x0
+        df = f(x, *args) - f0
+        J = df / dx
+    else:
+        J = []
+        for i in range(x0.size):
+            x = np.copy(x0)
```

Comment:
The target shape of `J` is known, so you can preallocate J and fill the rows. This is likely faster.